File: Top_100_Liked_Questions/53._Maximum_Subarray/solution.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        # we can keep track of the max sum seen so far while looping
        # if the current sum is negetive start fresh
        # because it will only bring down the combined sum
        # if positive, keep it and continue
        currSum = maxSum = float(-inf)
        
        for n in nums:
            if currSum > 0:
                currSum += n
            else:
                currSum = n
                
            maxSum = max(currSum, maxSum)
        
        return maxSum

File: Top_100_Liked_Questions/53._Maximum_Subarray/solution.java
class Solution {
    public int maxSubArray(int[] nums) {
        // one pass algorithm
        int ans = Integer.MIN_VALUE;
        int sum = 0;
        for (int n : nums) {
            // if the sum is negative, start fresh
            if (sum < 0)
                sum = 0;

            sum += n;
            ans = Math.max(ans, sum);
        }

        return ans;
    }
}